packages/styled-components/src/types.ts

```diff
@@ -77,7 +77,7 @@ export type Attrs<Props extends object> =
   | (ExecutionProps & Props)
   | ((props: ExecutionContext & Props) => Partial<Props>);
 
-export type RuleSet<Props extends object> = Interpolation<ExecutionContext & Props>[];
```

Comment:
Thanks @probablyup. Are the tests excluded from `tsc` intentionally? Seems like we'd want to include them.

Reply:
Looks like the problem is actually 47c370fa4ddfd29b07022491181320956755dac7

Reply:
Okay, got it figured out! The problem is that `Interpolation` can't disambiguate between `StyledFunction` and `IStyledComponent` because they both have function signatures. `p => p.color` could match either one; we're really depending on the idea that there's only one kind of callable in an interpolation.

Will add another commit to this PR